woocommerce/woocommerce-android
# [Better Errors] Introduce troubleshooting dialog

Summary
==========
Fix issue #10797 by introducing the troubleshooting dialog once the Timeout Retry fails.

Screenshots
==========
<img src="https://github.com/woocommerce/woocommerce-android/assets/5920403/d9ffc569-e8af-4f75-8485-7dd0db65e293" width="270" height="570" />

How to Test
==========
## Test setup

### Net Limiters
First, install the MacOS `Network Link Conditioner` software from the [Xcode Additional Tools](https://developer.apple.com/download/all/). This software, once installed, will be available in your Mac settings and allow you to configure network limiters that will make it easier to reproduce Timeout scenarios inside an emulator.

I suggest using a custom network configuration with a very restricted network. This way, we can ensure that the device still detects a connection, but it takes forever to finish any request.

<img width="878" alt="Screenshot 2024-02-28 at 23 12 14" src="https://github.com/woocommerce/woocommerce-android/assets/5920403/dbf24c2b-4917-41b4-b71e-504cb3dae77e">
<img width="676" alt="Screenshot 2024-02-28 at 23 12 57" src="https://github.com/woocommerce/woocommerce-android/assets/5920403/5d646c0d-24b9-40c7-b257-39bd1864133e">

### Code changes
Since the timeout error is hard to reproduce, and this test scenario requires two timeout events one after the other, I suggest applying this patch for better testing:

``` text
Subject: [PATCH] Fix lint issues
---
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt	(revision 4e10f066c0fd275eecb1674004715177c97d22bf)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt	(date 1709594034692)
@@ -407,27 +407,10 @@
         pagedListWrapper.listError
             .filter { !dismissListErrors }
             .filterNotNull()
-            .observe(this) { error ->
-                when (error.type) {
-                    PARSE_ERROR -> {
-                        viewState = viewState.copy(
-                            isErrorFetchingDataBannerVisible = true,
-                            isSimplePaymentsAndOrderCreationFeedbackVisible = false
-                        )
-                    }
-                    TIMEOUT_ERROR -> {
-                        when {
-                            shouldRetry && noTimeoutHappened -> {
-                                triggerEvent(RetryLoadingOrders)
-                            }
-                            else -> viewState = viewState.copy(
-                                shouldDisplayTroubleshootingBanner = true
-                            )
-                        }
-                        noTimeoutHappened = false
-                    }
-                    else -> triggerEvent(ShowErrorSnack(R.string.orderlist_error_fetch_generic))
-                }
+            .observe(this) { _ ->
+                viewState = viewState.copy(
+                    shouldDisplayTroubleshootingBanner = true
+                )
             }
         this.activePagedListWrapper = pagedListWrapper
```

## Test scenario

1. Open the app, go to the Orders section
2. Once inside the Orders list, wait for the Timeout error to happen first
3. Verify the Troubleshooting dialog is displayed
4. Check the `Contact Support` button works as expected
5. Check the `troubleshooting` button does nothing (it will be introduced in https://github.com/woocommerce/woocommerce-android/pull/11001)

Update release notes:
